File: src/main/java/com/fifo/ticketing/domain/user/entity/OAuthProvider.java
package com.fifo.ticketing.domain.user.entity;

import com.fifo.ticketing.domain.user.dto.oauth.UserOAuthDetails;
import com.fifo.ticketing.global.exception.ErrorCode;
import com.fifo.ticketing.global.exception.ErrorException;
import java.util.Map;
import org.springframework.security.oauth2.core.user.OAuth2User;

@SuppressWarnings("unchecked")
public enum OAuthProvider {

    GOOGLE {
        @Override
        public UserOAuthDetails extract(OAuth2User oAuth2User) {
            Map<String, Object> attributes = oAuth2User.getAttributes();
            return new UserOAuthDetails(
                attributes.get("name").toString(),
                attributes.get("email").toString(),
                attributes,
                Role.USER
            );
        }
    },
    NAVER {
        @Override
        public UserOAuthDetails extract(OAuth2User oAuth2User) {
            Map<String, Object> attributes = oAuth2User.getAttributes();
            Map<String, Object> properties = (Map<String, Object>) attributes.get("response");
            return new UserOAuthDetails(
                properties.get("name").toString(),
                properties.get("id") + "@naver.com",
                properties,
                Role.USER
            );
        }
    },
    KAKAO {
        @Override
        public UserOAuthDetails extract(OAuth2User oAuth2User) {
            Map<String, Object> attributes = oAuth2User.getAttributes();
            Map<String, Object> properties = (Map<String, Object>) attributes.get("properties");
            return new UserOAuthDetails(
                properties.get("nickname").toString(),
                attributes.get("id") + "@kakao.com",
                properties,
                Role.USER
            );
        }
    };

    public static OAuthProvider from(String provider) {
        try {
            return OAuthProvider.valueOf(provider.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new ErrorException(ErrorCode.NOT_FOUND_PROVIDER);
        }
    }

    public abstract UserOAuthDetails extract(OAuth2User oAuth2User);
}

File: src/main/java/com/fifo/ticketing/domain/user/entity/OAuthProvider.kt
package com.fifo.ticketing.domain.user.entity

import com.fifo.ticketing.domain.user.dto.oauth.UserOAuthDetails
import com.fifo.ticketing.global.exception.ErrorCode
import com.fifo.ticketing.global.exception.ErrorException
import org.springframework.security.oauth2.core.user.OAuth2User
import java.util.*

enum class OAuthProvider {
    GOOGLE {
        override fun extract(oAuth2User: OAuth2User): UserOAuthDetails {
            val attributes = oAuth2User.attributes
            return UserOAuthDetails(
                attributes["name"].toString(),
                attributes["email"].toString(),
                attributes,
                Role.USER
            )
        }
    },
    NAVER {
        override fun extract(oAuth2User: OAuth2User): UserOAuthDetails {
            val attributes = oAuth2User.attributes
            val properties = attributes["response"] as Map<String, Any>?
            return UserOAuthDetails(
                properties!!["name"].toString(),
                properties["id"].toString() + "@naver.com",
                properties,
                Role.USER
            )
        }
    },
    KAKAO {
        override fun extract(oAuth2User: OAuth2User): UserOAuthDetails {
            val attributes = oAuth2User.attributes
            val properties = attributes["properties"] as Map<String, Any>?
            return UserOAuthDetails(
                properties!!["nickname"].toString(),
                attributes["id"].toString() + "@kakao.com",
                properties,
                Role.USER
            )
        }
    };

    abstract fun extract(oAuth2User: OAuth2User): UserOAuthDetails

    companion object {
        fun from(provider: String): OAuthProvider {
            try {
                return valueOf(provider.uppercase(Locale.getDefault()))
            } catch (e: IllegalArgumentException) {
                throw ErrorException(ErrorCode.NOT_FOUND_PROVIDER)
            }
        }
    }
}
